src/lib/HST/Effect/Report.hs

```diff
@@ -54,13 +58,161 @@ import           HST.Effect.Cancel ( Cancel, cancel, runCancel )
 data Severity = Internal | Error | Warning | Info | Debug
  deriving ( Show, Eq )
 
--- | A messages that can be 'report'ed.
-data Message = Message { msgSeverity :: Severity, msgText :: String }
+-- | A message that can be 'report'ed.
+data Message = Message { msgSeverity :: Severity
+                       , msgSrcSpan  :: Maybe S.MsgSrcSpan
+                       , msgText     :: String
+                       }
  deriving ( Show, Eq )
 
+-- | Like using the 'Message' constructor directly, but takes a 'S.SrcSpan'
+--   instead of a @Maybe@ 'S.MsgSrcSpan' and does the conversion.
+message :: Severity -> S.SrcSpan a -> String -> Message
+message severity srcSpan = Message severity (S.toMsgSrcSpan srcSpan)
+
 -- TODO Add @Pretty@ instance for messages.
+-- | Pretty-prints a message without a code excerpt.
+--
+--   Only used for testing purposes where the 'HST.Effect.InputFile' effect is
+--   not always handled.
 showPrettyMessage :: Message -> String
-showPrettyMessage (Message severity msg) = show severity ++ ": " ++ msg
+showPrettyMessage (Message severity Nothing msg)
+  = show severity ++ ": " ++ msg
+showPrettyMessage (Message severity (Just src) msg)
+  = show severity ++ ": " ++ msg ++ "\n  In " ++ prettyMsgSrcSpan src ++ "."
+
+-- | Pretty-prints a message with a code excerpt.
+showPrettyMessageWithExcerpt :: Member InputFile r => Message -> Sem r String
+showPrettyMessageWithExcerpt (Message severity srcSpan msg) = do
+  excerpt <- displayCodeExcerpt srcSpan
+  return
+    $ show severity
+    ++ ": "
+    ++ msg
+    ++ if null excerpt then "" else '\n' : excerpt
+
+-- | Builds a string displaying the file path and the line and column of the
+--   start and end of the given message source span.
+prettyMsgSrcSpan :: S.MsgSrcSpan -> String
+prettyMsgSrcSpan src
+  = S.msgSrcSpanFilePath src ++ ':' : prettyMsgSrcSpanNumbers src
+
+-- | Builds a string displaying the line and column of the given message source
+--   span start and end.
+prettyMsgSrcSpanNumbers :: S.MsgSrcSpan -> String
+prettyMsgSrcSpanNumbers src = show (S.msgSrcSpanStartLine src)
+  ++ ':'
+  : show (S.msgSrcSpanStartColumn src)
+  ++ '-'
+  : show (S.msgSrcSpanEndLine src) ++ ':' : show (S.msgSrcSpanEndColumn src)
+
+-- TODO Should the following function be in its own module? It can't be put
+-- back into 'HST.Frontend.Transformer.Messages' because of cyclic imports.
+-- | Displays an excerpt of the input code specified by the given source span.
+--
+--   The excerpt consists of an introductory line with the file path and the
+--   source span numbers, all lines of the input program that are at least
+--   partially contained in the given source span, including their line
+--   numbers, and marks showing the start and end of the spanned code.
+displayCodeExcerpt :: Member InputFile r => Maybe S.MsgSrcSpan -> Sem r String
```

Comment:
In my tests, I found out that `displayCodeExcerpt` can also display the following `SrcSpan` that is invalid.
`S.MsgSrcSpan "./example/ExampleQueue1.hs" 1 4 1 3`.

Reply:
This should be fixed now, but you should run your tests again to be sure that it's actually fixed and that I haven't introduced another bug.